clients/pubsub-common/src/main/java/org/eclipse/hono/client/pubsub/subscriber/PubSubSubscriberFactory.java

```diff
@@ -0,0 +1,68 @@
+/**
+ * Copyright (c) 2023 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.eclipse.hono.client.pubsub.subscriber;
+
+import java.util.Optional;
+
+import com.google.cloud.pubsub.v1.MessageReceiver;
+
+import io.vertx.core.Future;
+
+/**
+ * A factory for creating Pub/Sub subscribers scoped to a Google Cloud Project.
+ */
+public interface PubSubSubscriberFactory {
+
+    /**
+     * Closes the subscriber with the given topic if it exists.
+     * <p>
+     * This method is expected to be invoked as soon as the subscriber is no longer needed.
+     *
+     * @param subscription The subscription of the subscriber to remove.
+     * @param prefix The prefix of the topic of the subscriber to remove, e.g. the tenantId.
+     * @return A future that is completed when the close operation completed or a succeeded future if no subscriber
+     *         existed with the given topic.
+     */
+    Future<Void> closeSubscriber(String subscription, String prefix);
+
+    /**
+     * Closes all cached subscriber. This method is expected to be invoked especially before the application shuts down.
+     *
+     * @return A future that is succeeded when all subscriber are closed or a failed future if any subscriber can not be
+     *         closed.
+     */
+    Future<Void> closeAllSubscriber();
+
+    /**
+     * Gets a subscriber for receiving data from Pub/Sub.
+     * <p>
+     * The subscriber returned may be either newly created or it may be an existing subscriber for the given
+     * subscription.
```

Comment:
I don’t think so, but this won’t be the case. This method should (and will) be only called with a MessageReceiver belonging to this subscription name.
More precisely: There will be 3 classes which calls the getOrCreateSuscriber() method to get a PubSubSubscriberClient for a specific subscription:
- The Internal Command Consumer (which consumes messages on the internal subscription _adapterInstanceId.command_internal_ based on the topic _adapterInstanceId.command_internal_)
- The Command Consumer Factory (which consumes messages on the command subscription _tenantId.command_ based on the topic _tenantId.command_)
- The Notification Receiver (which consumes messages on the notification subscriptions _registry-device.notification_ and _registry-tenant.notification_ based on the topics _registry-device.notification_ and _registry-tenant.notification_)

Each class contains its own MessageReceiver and subscribes to **its** subscription. The idea is to share the subscriptions inside the class, so they don’t have to be created again and again, because creating the subscription can take some time. The subscription name and also the message receiver should be of course the same for one subscription. So the subscription won’t be shared with multiple MessageReceivers.